Remotion/Web/ClientScript/Scripts/SmartPage.ts

```diff
@@ -0,0 +1,1307 @@
+    var isTriggeredByCurrentSubmitter = this._submitState.Submitter === this.GetActiveElement();
+    var hasQueuedSubmit = this._submitState.NextSubmitState != null;
+    if (!hasQueuedSubmit && isAsyncAutoPostback && !isTriggeredByCurrentSubmitter)
+      return true;
+
+    this.ShowStatusIsSubmittingMessage();
+    return false;
+  };
+
+  // Event handler for Window.OnScroll.
+  public OnScroll(): void
+  {
+    if (this._statusMessageWindow != null)
+      this.AlignStatusMessage(this._statusMessageWindow);
+    this.ExecuteEventHandlers('onscroll');
+  };
+
+  // Event handler for Window.OnResize.
+  public OnResize(): void
+  {
+    if (this._statusMessageWindow != null)
+      this.AlignStatusMessage(this._statusMessageWindow);
+    this.ExecuteEventHandlers('onresize');
+  };
+
+  // Sends an AJAX request to the server.
+  // successHandler: function (args { Status }), called when the request succeeds
+  // errorHandler: function (args { Status }), called when the request fails
+  public SendOutOfBandRequest (url: string, successHandler: OutOfBandRequestSuccessHandler, errorHandler: OutOfBandRequestErrorHandler): void
+  {
+    ArgumentUtility.CheckNotNullAndTypeIsString('url', url);
+    ArgumentUtility.CheckNotNullAndTypeIsFunction('successHandler', successHandler);
+    ArgumentUtility.CheckNotNullAndTypeIsFunction('errorHandler', errorHandler);
+
+    if ((this._isOnBeforeUnloadExecuting || this._isOnUnloadExecuting) && navigator.sendBeacon)
+    {
+      var result = navigator.sendBeacon (url, null);
+      if (result)
+      {
+        let args = { Status : 200 };
+        successHandler (args);
+      }
+      else
+      {
+        let args = { Status: 400 };
+        errorHandler (args);
+      }
+    }
+    else
+    {
+      var xhr = new XMLHttpRequest();
+
+      var readStateDone = 4;
+      var httpStatusSuccess = 299;
+      var method = 'GET';
+      var isAsyncCall = true;
+
+      xhr.open (method, url, isAsyncCall);
+      xhr.onreadystatechange = function ()
+      {
+        if (this.readyState === readStateDone)
+        {
+          var args = { Status : this.status };
+          if (this.status > 0 && this.status <= httpStatusSuccess)
+            successHandler (args);
+          else
+            errorHandler (args);
+        }
+      };
+      xhr.send();
+    }
+  };
+
+  private AddEventHandler(object: string | {}, eventType: string, handler: (eventObject: JQueryEventObject) => any): void
+  {
+    $(object).bind(eventType, handler);
+    return;
+  };
+
+  private RemoveEventHandler(object: string | {}, eventType: string, handler: (eventObject: JQueryEventObject) => any): void
+  {
+    $(object).unbind(eventType, handler);
+    return;
+  }
+
+  // Executes the event handlers.
+  // eventHandlers: an array of event handlers.
+  private ExecuteEventHandlers<TEventName extends SmartPageEvent>(eventName: TEventName, ...args: Parameters<SmartPageEventMap[TEventName]>)
```

Comment:
I did this change inline instead because it is necessary to make sure that the method can be type-checked. Otherwise it would be necessary to cast any usage to the correct type, which makes all callsites more complex.

Reply:
i.c.

Reply:
This could also be a nice TS-syntax crazyness topic for a DevJf. After all, this function probably contains more non-letter characters than it does letter characters :D